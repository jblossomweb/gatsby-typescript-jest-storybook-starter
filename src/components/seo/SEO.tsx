import React from "react"
import PropTypes from "prop-types"
import { Helmet, HelmetProps } from "react-helmet"

export interface Meta {
  title?: string
  description?: string
  author?: string
  [key: string]: string | undefined
}

export interface Props {
  title: string
  description?: string
  author?: string
  lang?: string
  meta?: HelmetProps["meta"] | Meta[] | any[]
}

const SEO: React.FC<Props> = ({ title, description, author, lang, meta }) => {
  return (
    <Helmet
      htmlAttributes={{ lang }}
      title={title}
      meta={[
        {
          name: `description`,
          content: description,
        },
        {
          property: `og:title`,
          content: title,
        },
        {
          property: `og:description`,
          content: description,
        },
        {
          property: `og:type`,
          content: `website`,
        },
        {
          name: `twitter:card`,
          content: `summary`,
        },
        {
          name: `twitter:creator`,
          content: author,
        },
        {
          name: `twitter:title`,
          content: title,
        },
        {
          name: `twitter:description`,
          content: description,
        },
      ].concat(meta as any[])}
    />
  )
}

SEO.defaultProps = {
  lang: `en`,
  meta: [],
  description: ``,
}

SEO.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string,
  author: PropTypes.string,
  lang: PropTypes.string,
  meta: PropTypes.arrayOf(PropTypes.object),
}

export default SEO
