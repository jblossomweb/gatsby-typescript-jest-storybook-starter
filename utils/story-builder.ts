import * as knobs from "@storybook/addon-knobs"
import { storiesOf } from "@storybook/react"

const { withKnobs } = knobs

export interface KnobsInterface {
  text: (name: string, val: string) => string
  number: (name: string, val: number) => number
  date: (name: string, val: number) => number
  boolean: (name: string, val: boolean) => boolean
  object: (name: string, val: any) => any
  select: (name: string, options: any, val: any) => any
}

export type Story = (...args: any[]) => JSX.Element

export interface Stories {
  [key: string]: Story
}

const storyBuilder = (stories: Stories, storyPath: string) => {
  const builtStories = storiesOf(storyPath, module)
  builtStories.addDecorator(withKnobs as any)

  Object.keys(stories).forEach(key => {
    builtStories.add(key, () => stories[key](knobs))
  })
}

export default storyBuilder
