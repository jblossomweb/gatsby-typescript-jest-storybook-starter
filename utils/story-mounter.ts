import { Stories } from "./story-builder"
import "./enzyme-adapter"
import { shallow, mount, ReactWrapper } from "enzyme"
import mockKnobs from "../__mocks__/storybook-knobs"

export interface MountedStory {
  wrapper: ReactWrapper
  component: ReactWrapper
  props: any
  elements: { [key: string]: ReactWrapper }
}

const storyMounter = (
  stories: Stories,
  mainComponent: React.Component<any> | React.FC<any>,
  getElements?: (component: ReactWrapper) => MountedStory["elements"],
  mountShallow?: boolean
) => {
  const mountedStories: { [key: string]: MountedStory } = {}
  // tslint:disable-next-line: ban-types
  const mounter: Function = mountShallow ? shallow : mount
  Object.keys(stories).forEach((key: string) => {
    const story = stories[key]
    const wrapper: ReactWrapper = mounter(story(mockKnobs))
    const component: ReactWrapper = wrapper.find(mainComponent)
    const props = component.props()
    const elements = getElements ? getElements(component) : {}
    mountedStories[key] = {
      wrapper,
      component,
      props,
      elements,
    }
  })
  return mountedStories
}

export default storyMounter
