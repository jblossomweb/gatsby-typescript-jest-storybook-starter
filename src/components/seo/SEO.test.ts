import { Helmet } from "react-helmet"
import storyMounter from "../../../utils/story-mounter"
import { renderTree } from "../../../utils/story-snapshot-tests"
import { stories } from "./SEO.stories"
import SEO from "./SEO"

describe("SEO", () => {
  const mountedStories = storyMounter(stories, SEO, component => ({
    helmet: component.find(Helmet),
  }))

  Object.keys(mountedStories).forEach(key => {
    describe(key, () => {
      const story = stories[key]
      const mountedStory = mountedStories[key]
      it("mounts the SEO component", () => {
        expect(mountedStory.component.length).toBe(1)
      })

      it("mounts the Helmet component", () => {
        const { helmet } = mountedStory.elements
        expect(helmet.length).toBe(1)
      })

      it("passes props.lang to Helmet htmlAttributes", () => {
        const { helmet } = mountedStory.elements
        const { lang } = mountedStory.props
        expect(helmet.prop("htmlAttributes")).toEqual({ lang })
      })

      it("passes props.title to Helmet title", () => {
        const { helmet } = mountedStory.elements
        const { title } = mountedStory.props
        expect(helmet.prop("title")).toEqual(title)
      })

      it("generates and appends expected data to Helmet meta", () => {
        const { helmet } = mountedStory.elements
        const { description, title, author, meta } = mountedStory.props
        const expected = [
          {
            name: `description`,
            content: description,
          },
          {
            property: `og:title`,
            content: title,
          },
          {
            property: `og:description`,
            content: description,
          },
          {
            property: `og:type`,
            content: `website`,
          },
          {
            name: `twitter:card`,
            content: `summary`,
          },
          {
            name: `twitter:creator`,
            content: author,
          },
          {
            name: `twitter:title`,
            content: title,
          },
          {
            name: `twitter:description`,
            content: description,
          },
        ].concat(meta as any[])
        expect(helmet.prop("meta")).toEqual(expected)
      })

      it("matches empty DOM snapshot", () => {
        const tree = renderTree(story)
        expect(tree).toMatchSnapshot()
      })
    })
  })
})
