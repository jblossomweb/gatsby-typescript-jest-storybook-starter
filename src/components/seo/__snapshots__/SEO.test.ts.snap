// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEO author matches empty DOM snapshot 1`] = `null`;

exports[`SEO description matches empty DOM snapshot 1`] = `null`;

exports[`SEO sample matches empty DOM snapshot 1`] = `null`;
